iree/compiler/Codegen/LLVMGPU/LLVMGPUSwizzle.cpp

```diff
@@ -0,0 +1,112 @@
+// Copyright 2022 The IREE Authors
+//
+// Licensed under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+#include "iree/compiler/Codegen/PassDetail.h"
+#include "iree/compiler/Codegen/Passes.h"
+
+namespace mlir {
+namespace iree_compiler {
+
+/// This function implements the following swizzling logic
+/// void getTiledId2(unsigned x, unsigned y, unsigned* tiledx,
+///                 unsigned* tiledy) {
+///  unsigned t_tiledx = (x + (y % tile) * grid_size_x) / tile;
+///  unsigned t_tiledy = (y / tile) * tile +
+///      (x + (y % tile) * grid_size_x) % tile;
+///  bool c = grid_size_y % tile != 0 &&
+///      ((y / tile) * tile + tile) > grid_size_y;
+///  *tiledx = c ? x : t_tiledx;
+///  *tiledy = c ? y : t_tiledy;
+/// }
+static void makeSwizzledId(Location loc, OpBuilder b, Value workgroupIdX,
+                           Value workgroupIdY, Value gridSizeX, Value gridSizeY,
+                           Value& SwizzledIdX, Value& SwizzledIdY,
+                           unsigned swizzleTile) {
+  Value zero = b.create<arith::ConstantIndexOp>(loc, 0);
+  Value tile = b.create<arith::ConstantIndexOp>(loc, swizzleTile);
+  Value yModTile = b.create<arith::RemUIOp>(loc, workgroupIdY, tile);
+  Value yDivTile = b.create<arith::DivUIOp>(loc, workgroupIdY, tile);
+  Value swizzleParam = b.create<arith::MulIOp>(loc, yModTile, gridSizeX);
+  Value swizzleParam2 =
+      b.create<arith::AddIOp>(loc, workgroupIdX, swizzleParam);
+  Value swizzleParam3 = b.create<arith::RemUIOp>(loc, swizzleParam2, tile);
+  Value swizzleParam4 = b.create<arith::MulIOp>(loc, yDivTile, tile);
+  Value unboundedSwizzledIdX =
+      b.create<arith::DivUIOp>(loc, swizzleParam2, tile);
+  Value unboundedSwizzledIdY =
+      b.create<arith::AddIOp>(loc, swizzleParam3, swizzleParam4);
+  Value gyModTile = b.create<arith::RemUIOp>(loc, gridSizeY, tile);
+  Value gyAddTile = b.create<arith::AddIOp>(loc, swizzleParam4, tile);
+  Value condition1 =
+      b.create<arith::CmpIOp>(loc, arith::CmpIPredicate::ne, gyModTile, zero);
+  Value condition2 = b.create<arith::CmpIOp>(loc, arith::CmpIPredicate::ugt,
+                                             gyAddTile, gridSizeY);
+  Value condition3 = b.create<arith::AndIOp>(loc, condition1, condition2);
+  SwizzledIdX = b.create<arith::SelectOp>(loc, condition3, workgroupIdX,
+                                          unboundedSwizzledIdX);
+  SwizzledIdY = b.create<arith::SelectOp>(loc, condition3, workgroupIdY,
+                                          unboundedSwizzledIdY);
+}
+namespace {
+struct LLVMGPUSwizzlePass : public LLVMGPUSwizzleBase<LLVMGPUSwizzlePass> {
```

Comment:
this is the kind of stuff we'd want on Vulkan and CPU as well, so even if staging it here under LLVMGPU it'd be good to ensure this code stays generic and portable to Codegen/Common/ so that it can be reused.

Reply:
Makes sense, I can move the pass to Codegen/Common/. This pass can work on any IR after tile and distribute phase. So there is nothing CUDA specific.

Reply:
+1. I'm gonna pick this up some days later once #8425 lands. ;-P